### PR TITLE
issue #119: use subject instead of rereading file when adding error context

### PR DESCRIPTION
**Pull request by mascarenhas**

### CODE REVIEW

File: typedlua/tlchecker.lua [hunk 1]
```diff
@@ -1962,9 +1962,14 @@ function tlchecker.typecheck (ast, subject, filename, strict, integer, color)
...
-local function get_source_line(filename, l)
+local function get_source_line(subject, filename, l)
+  if not subject then
+    local file = io.open(filename)
+    subject = file:read("*a")
+    file:close()
+  end
   local i = 1
-  for source_line in io.lines(filename) do
+  for source_line in subject:gmatch("[^\r\n]*") do
```
Comment (daurnimator) [inline typedlua/tlchecker.lua]: I think this will result in skipping empty lines and getting the count wrong?

**Review by daurnimator — CHANGES_REQUESTED** [review]

File: typedlua/tlchecker.lua (hunk 1 above)
Comment (daurnimator) [reply typedlua/tlchecker.lua]: In Lua < 5.3.4 (yes it was changed in a minor version!) gmatch will return extra 0 length matches.

See e.g. 
```lua
for thing in ("a\nb\n\nfoo"):gmatch("[^\r\n]*") do print(thing) end
```

Comment (daurnimator) [reply typedlua/tlchecker.lua]: It was undefined behaviour that finally got changed to the "expected" behaviour in the 5.3.3 release. I believe for 5.4 the behaviour will actually be documented (and hence relied on).

> I guess I will have to write a custom iterator using string.byte...

Just modify your gmatch slightly. `^(.-)\r?\n` should work I think? (excluding the final line of the file).
Though you have lpeg as a dependency already! Make something nicer :)

Comment (mascarenhas) [reply typedlua/tlchecker.lua]: Using lpeg for that is shooting a fly with a bazooka. :-) I just did a loop using `string.find` looking for `\n`.

File: typedlua/tlchecker.lua [hunk 2]
```diff
@@ -1962,14 +1962,25 @@ function tlchecker.typecheck (ast, subject, filename, strict, integer, color)
...
+  if not subject then
+    local file = io.open(filename)
+    subject = file:read("*a")
+    file:close()
+  end
   local i = 1
-  for source_line in io.lines(filename) do
+  repeat
+    local lstart = 1
+    local lend = subject:find("\n", lstart) or #subject
```
Comment (daurnimator) [inline typedlua/tlchecker.lua]: Use the `plain` argument to find for much more speed.

Comment (daurnimator) [reply typedlua/tlchecker.lua]: (or have the pattern strip `\r`)

File: typedlua/tlchecker.lua [hunk 3]
```diff
@@ -1962,14 +1962,25 @@ function tlchecker.typecheck (ast, subject, filename, strict, integer, color)
...
     if i == l then
-      return (string.gsub(source_line, "\t", " "))
+      local line = subject:sub(lstart, lend or #subject)
+      line = line:gsub("[\r\n]", "")
```
Comment (daurnimator) [inline typedlua/tlchecker.lua]: this shouldn't be necessary if you have the pattern strip `\r`

Comment (daurnimator) [reply typedlua/tlchecker.lua]: you may want to strip/escape other control characters.

File: typedlua/tlchecker.lua [hunk 4]
```diff
@@ -1962,14 +1962,25 @@ function tlchecker.typecheck (ast, subject, filename, strict, integer, color)
   return env.messages
 end
 
-local function get_source_line(filename, l)
+local function get_source_line(subject, filename, l)
+  if not subject then
+    local file = io.open(filename)
+    subject = file:read("*a")
```
Comment (daurnimator) [inline typedlua/tlchecker.lua]: `assert` this (reads can fail)

**Review by daurnimator — CHANGES_REQUESTED** [review]

File: typedlua/tlchecker.lua [hunk 5]
```diff
@@ -1964,23 +1964,21 @@ end
...
-    local file = io.open(filename)
+    local file = assert(io.open(filename))
     subject = file:read("*a")
     file:close()
   end
   local i = 1
   repeat
     local lstart = 1
-    local lend = subject:find("\n", lstart) or #subject
+    local lend, lnext = subject:find("\r?\n", lstart) or #subject
```
[the remark this answers is not in the document]
Comment (mascarenhas) [reply typedlua/tlchecker.lua]: I was staring at a failing test case the whole time and not really seeing it. It is ok now.

File: typedlua/tlchecker.lua (hunk 3 above)
Comment (daurnimator) [reply typedlua/tlchecker.lua]: Even so: we don't want to print random control characters to the user's screen.

File: typedlua/tlchecker.lua [hunk 6]
```diff
@@ -1962,14 +1962,23 @@ function tlchecker.typecheck (ast, subject, filename, strict, integer, color)
   return env.messages
 end
 
-local function get_source_line(filename, l)
+local function get_source_line(subject, filename, l)
+  if not subject then
+    local file = assert(io.open(filename))
+    subject = file:read("*a")
```
Comment (daurnimator) [inline typedlua/tlchecker.lua]: Still need an `assert` (previous comment dropped off)

File: typedlua/tlchecker.lua [hunk 7]
```diff
@@ -1962,14 +1962,23 @@ function tlchecker.typecheck (ast, subject, filename, strict, integer, color)
...
+  end
   local i = 1
-  for source_line in io.lines(filename) do
+  local lstart = 1
+  repeat
+    local lend, lnext = subject:find("\r?\n", lstart)
     if i == l then
-      return (string.gsub(source_line, "\t", " "))
+      local line = subject:sub(lstart, lend and lend-1 or #subject)
+      return (line:gsub("\t", " "))
```
Comment (daurnimator) [inline typedlua/tlchecker.lua]: (previous comment dropped off): escape or strip all non-printable characters

**Review by daurnimator — CHANGES_REQUESTED** [review]

File: typedlua/tlchecker.lua (hunk 7 above)
Comment (mascarenhas) [reply typedlua/tlchecker.lua]: I saw the comment, I think this was done as a hack to fix placement of the arrow when tabs are involved, but that placement will already be broken in several other situations, such as with any line that has multibyte characters.

**Pull request merged**